src/sql/src/func.rs

```diff
@@ -143,6 +124,43 @@ impl TypeCategory {
         }
     }
 
+    /// Like `from_type`, but for catalog types.
+    // TODO(benesch): would be nice to figure out how to share code with
+    // `from_type`, but the refactor to enable that would be substantial.
+    pub fn from_catalog_type<T>(catalog_type: &CatalogType<T>) -> Self
+    where
+        T: TypeReference,
+    {
+        // Keep this in sync with `from_type`.
+        match catalog_type {
+            CatalogType::Array { .. } | CatalogType::Int2Vector => Self::Array,
+            CatalogType::Bool => Self::Boolean,
+            CatalogType::Bytes | CatalogType::Jsonb | CatalogType::Uuid => Self::UserDefined,
+            CatalogType::Date
+            | CatalogType::Time
+            | CatalogType::Timestamp
+            | CatalogType::TimestampTz => Self::DateTime,
+            CatalogType::Float32
+            | CatalogType::Float64
+            | CatalogType::Int16
+            | CatalogType::Int32
+            | CatalogType::Int64
+            | CatalogType::Oid
+            | CatalogType::RegClass
+            | CatalogType::RegProc
+            | CatalogType::RegType
+            | CatalogType::Numeric { .. } => Self::Numeric,
+            CatalogType::Interval => Self::Timespan,
+            CatalogType::List { .. } => Self::List,
+            CatalogType::PgLegacyChar
+            | CatalogType::String
+            | CatalogType::Char { .. }
+            | CatalogType::VarChar { .. } => Self::String,
+            CatalogType::Record { .. } => TypeCategory::Composite,
+            CatalogType::Map { .. } | CatalogType::Pseudo => Self::Pseudo,
```

Comment:
Why Map is a pseudo-type?

Reply:
This is just a transliteration of the existing type categories. Map should likely be its own type, _a la_ list.